public/js/notifications.js

```diff
@@ -0,0 +1,433 @@
+
+            // check for required API's
+            this.logger.debug(this.prefix + "checking for the required APIs and permissions");
+
+            let isValidated = true;
+            if (! ('ServiceWorker' in self)) {
+                this.logger.error(
+                    this.prefix
+                    + "this browser does not support the 'Service Worker API' in the"
+                    + " current context"
+                );
+                isValidated = false;
+            }
+            if (! ('Navigator' in self)) {
+                this.logger.error(
+                    this.prefix
+                    + "this browser does not support the 'Navigator API' in the"
+                    + " current context"
+                );
+                isValidated = false;
+            }
+            if (! ('Notification' in self)) {
+                this.logger.error(
+                    this.prefix
+                    + "this browser does not support the 'Notification API' in the"
+                    + " current context"
+                );
+                isValidated = false;
+            }
+            if (! isValidated) {
+                // we only log the error and exit early as throwing would completely hang up the web application
+                this.logger.error("The 'Notification' module is missing some required API's");
+                return;
+            }
+
+            this.logger.debug(this.prefix + "spawned");
+            this.load();
+        }
+
+        load () {
+            this.logger.debug(this.prefix + "loading");
+
+            // listen to render events on container for col1 (to inject notification toggle)
+            this.on('rendered', '#main > #col1.container', this.renderHandler, this);
+
+            // listen to controller (service worker) changes
+            navigator.serviceWorker.addEventListener('controllerchange', (event) => {
+                this.logger.debug(this.prefix + "new controller attached ", event.target.controller);
+                if (event.target.controller !== null) {
+                    // reset eventsource and handshake flag
+                    this.allowedToOperate = false;
+                    if (this.eventSource !== null && this.eventSource.readyState !== EventSource.CLOSED) {
+                        this.eventSource.close();
+                    }
+
+                    this.logger.debug(this.prefix + "send handshake to controller");
+                    event.target.controller.postMessage(
+                        JSON.stringify({
+                            command: 'handshake',
+                            version: VERSION
+                        })
+                    );
+                }
+            });
+
+            // listen to messages from the controller (service worker)
+            self.navigator.serviceWorker.addEventListener('message', (event) => {
+                if (event.data) {
```

Comment:
Please return earlier here.

Reply:
And make debug logs single-line if the line length is `> 120`.

Reply:
the line length is capped to a maximum of 120 in this file though